Add collection and favourite buttons to results screen

Layendan opened the pull request:
- Closes https://github.com/ppy/osu/discussions/28597

https://github.com/user-attachments/assets/9f6322f4-aad6-4f76-970b-a548f2f1d0c0

Layendan committed:
Add favourite button to results screen

Layendan committed:
Added collection button to result screen

Layendan committed:
Update favorite button

Layendan committed:
Add tests

Layendan committed:
Fix lint errors

Layendan committed:
Fix collection popover

peppy commented:
Please re-read the contribution guidelines (aka don't use `master` branch in the future).

You also made unnecessary changes to `osu.Desktop.slnf`. I've force-pushed these away.

peppy commented on osu.Game/Screens/Ranking/FavouriteButton.cs:
```diff
@@ -0,0 +1,164 @@
+
+namespace osu.Game.Screens.Ranking
+{
+    public partial class FavouriteButton : OsuAnimatedButton
```
There's now three versions of this code. This doesn't look to be copy-pasted from either other implementation. Did you miss these somehow? If so, please search for existing implementations before reimplementing.

If there's a reason for the different implementation, please let us know.

Layendan replied:
I originally copy pasted it from the `osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs` file, but that takes in an `APIBeatmapSet` object which doesn't exist on the rankings screen. 

The `BeatmapSet` version also doesn't handle signing into another account either (favorited on account A, but not on account B [or signs out -> signs back in]) since the page gets "reloaded" as the user signs out while the rankings screen does not get reloaded.

My original implementation in https://github.com/ppy/osu/commit/0bc14ba646e8fc7adfa9d8e3b53d0bf1341239e0 was more similar to the `BeatmapSet` version, but it had content shifting as the button loaded, and would load after a long time on slower internet; hence why I went forward with the reimplementation.

Layendan replied:
> The `BeatmapSet` version also doesn't handle signing into another account either (favorited on account A, but not on account B [or signs out -> signs back in]) since the page gets "reloaded" as the user signs out while the rankings screen does not get reloaded.


https://github.com/user-attachments/assets/35d68c78-fc22-4103-8df8-c552a4b9c22e

It also works if a guest user checks the rankings screen

peppy commented on osu.Game/Screens/Ranking/FavouriteButton.cs:
```diff
@@ -0,0 +1,164 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
Don't disable this.

peppy commented on osu.Game/Screens/Ranking/CollectionButton.cs:
```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
Don't disable this.

peppy commented on osu.Game.Tests/Visual/Ranking/TestSceneFavouriteButton.cs:
```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
Don't disable this.

Layendan commented:
> Please re-read the contribution guidelines (aka don't use `master` branch in the future).
I thought I went through everything, my bad. Do you want me to create a new fork using the correct branch and create a new pull request?

> You also made unnecessary changes to `osu.Desktop.slnf`. I've force-pushed these away.
I accidentally ran the formatter on the entire solution instead of just the files I changed. I thought I removed all of the unnecessary changes but missed this one.

Layendan committed:
Remove nullable disable

bdach commented on osu.Desktop.slnf:
this file been changed *again*

Layendan replied:
Fixing it atm

bdach commented:
also the history is now an utter mess. you don't do a merge after someone has force-pushed to your branch

Layendan commented:
> also the history is now an utter mess. you don't do a merge after someone has force-pushed to your branch, you pull and only _then_ continue working

I did pull before making changes, then committed and pushed. I don't know why VS made an automatic merge.

Should I go ahead on creating a new fork and pull request to clean up the history and fix forking from main branch?

bdach commented:
> Should I go ahead on creating a new fork and pull request to clean up the history and fix forking from main branch?

no that's not required.

reset your branch to https://github.com/ppy/osu/pull/28991/commits/6bb562db14ccd84ac27c45fe14623e9a443da7e4 and cherry-pick https://github.com/ppy/osu/pull/28991/commits/dec4047eea07deeaf8f0c9ead71679b1ec1f6114 onto that

Joehuu committed:
Inherit `GrayButton` instead

Also fixes hover highlight.

Joehuu committed:
Move some properties/bindables around

Joehuu committed:
Use same size button for everything

Joehuu committed:
Remove unnecessary `ReceivePositionalInputAt`

Results is not even using the new footer.

Joehuu committed:
Remove direct margin set in popover that was causing positioning offset

Joehuu committed:
Merge remote-tracking branch 'upstream/master' into pr/28991

Joehuu committed:
Remove unnecessary depth and padding set

Joehuu commented:
I have pushed some code quality changes.

Unsure how to fix the failing test.

Joehuu commented on osu.Game/Screens/Ranking/CollectionButton.cs:
```diff
@@ -0,0 +1,40 @@
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours)
+        {
+            Background.Colour = colours.Green;
```
Maybe this should be green only if the beatmap difficulty is in a collection?

Layendan replied:
I have it implemented in local, but I was wondering about the fade times and performance.

https://github.com/user-attachments/assets/c791cf7f-acc7-4cd7-ae01-e17f09f2071b

The video is using a 500ms duration (same as replay download button next to it) but I personally feel like it's not responsive enough and would rather change it to 300ms like [UIV2 rounded button](https://github.com/ppy/osu/blob/565107205c60f14488f1642011808d286d712c31/osu.Game/Graphics/UserInterfaceV2/RoundedButton.cs#L90)

300ms:

https://github.com/user-attachments/assets/a8660bcc-48bf-45a8-a2d4-d633a87a89aa


Also in terms of performance, should I be able to do this, or is this ill advised?
```cs
// Might have performance issues idk how the realm works
realm.Realm.RealmChanged += (_, _) => updateRealm();
```

```cs
private void updateRealm()
{
    current.Value = realm.Realm.All<BeatmapCollection>().AsEnumerable().Any(c => c.BeatmapMD5Hashes.Contains(beatmapInfo.MD5Hash));
}
```

Joehuu replied:
I'm not fluent in realm myself, so you'll have to wait for another dev.

frenzibyte replied:
seems @ppy/team-client is needed here.

smoogipoo replied:
No, you can't use `RealmChanged`. The way to do it is to use `RealmAccess.RegisterForNotifications()`, allowing you to register a callback. There's a few example usages around the place.

Layendan commented:
> Unsure how to fix the failing test.

I can't replicate [TestSceneLeadIn](https://github.com/ppy/osu/runs/28007589922#r0s0), but I found the issue with the others.

FavoriteButton makes a `GetBeatmapSetRequest` to the API, this is handled by the dummy api [here](https://github.com/ppy/osu/blob/aa2515763c8183ae83e971af56a2273ba2edef47/osu.Game/Tests/Visual/OnlinePlay/TestRoomRequestsHandler.cs#L199C17-L214C18); however, for some reason the realm freaks out at [this line](https://github.com/ppy/osu/blob/aa2515763c8183ae83e971af56a2273ba2edef47/osu.Game/Tests/Visual/OnlinePlay/TestRoomRequestsHandler.cs#L203C27-L203C110) when querying the db for existing beatmaps. 

I added the `HasFavourited` and `FavouriteCount` to the dummy api response which fixes a future issue, but does not fix the underlying issue. 

I did find [this issue](https://github.com/realm/realm-dotnet/issues/723) that seems similar to this problem and from what I can understand, [this line of code](https://github.com/ppy/osu/blob/aa2515763c8183ae83e971af56a2273ba2edef47/osu.Game/Tests/Visual/OnlinePlay/TestRoomRequestsHandler.cs#L203C60-L203C81) is not a valid query for realms. Changing it to `var baseBeatmap = beatmapManager.QueryBeatmap(b => b.OnlineID == getBeatmapSetRequest.ID);` fixes the issue although it means that the function of the dummy api is not correct.

Layendan committed:
Update collection state when users add/remove from collection

Layendan committed:
Fix tests

bdach commented on osu.Game/Tests/Visual/OnlinePlay/TestRoomRequestsHandler.cs:
```diff
@@ -197,9 +197,8 @@ public bool HandleRequest(APIRequest request, APIUser localUser, BeatmapManager
-                    var baseBeatmap = getBeatmapSetRequest.Type == BeatmapSetLookupType.BeatmapId
-                        ? beatmapManager.QueryBeatmap(b => b.OnlineID == getBeatmapSetRequest.ID)
-                        : beatmapManager.QueryBeatmap(b => b.BeatmapSet.OnlineID == getBeatmapSetRequest.ID);
+                    // Incorrect logic, see https://github.com/ppy/osu/pull/28991#issuecomment-2256721076 for reason why this change
+                    var baseBeatmap = beatmapManager.QueryBeatmap(b => b.OnlineID == getBeatmapSetRequest.ID);
```
This cannot be left in like this. If you're going to try a fix, then at least try something like

```csharp
                    var baseBeatmap = getBeatmapSetRequest.Type == BeatmapSetLookupType.BeatmapId
                        ? beatmapManager.QueryBeatmap(b => b.OnlineID == getBeatmapSetRequest.ID)
                        : beatmapManager.QueryBeatmapSet(s => s.OnlineID == getBeatmapSetRequest.ID)?.PerformRead(s => s.Beatmaps.First().Detach());
```

which at least has a semblancy of correctness and should not cause realm to die of death (although I didn't verify this, because I can't get the test failure to repro on my box, on a quick look).

bdach commented on osu.Game/Screens/Ranking/CollectionButton.cs:
```diff
@@ -0,0 +1,81 @@
+    public partial class CollectionButton : GrayButton, IHasPopover
+    {
+        private readonly BeatmapInfo beatmapInfo;
+        private readonly Bindable<bool> current;
```
rename to `isInAnyCollection`

bdach commented on osu.Game/Screens/Ranking/CollectionButton.cs:
```diff
@@ -0,0 +1,81 @@
+            collectionSubscription?.Dispose();
+        }
+
+        private void updateRealm(IRealmCollection<BeatmapCollection> sender, ChangeSet? changes)
```
rename to `collectionsChanged`

Layendan committed:
update var names and test logic

Layendan commented:
Does the pr look good, or should I make any other changes?

Joehuu committed:
Make collection button test less broken

Joehuu reviewed (dismissed):
Looks good to me.

bdach committed:
Merge branch 'master' into master

bdach reviewed (changes requested):
Got this crash when testing the log in/out logic of the favourite button:

```
Unhandled exception. osu.Framework.Graphics.Drawable+InvalidThreadForMutationException: Cannot mutate the Transforms on a Loaded Drawable while not on the update thread. Consider using Schedule to schedule the mutation operation.
   at osu.Framework.Graphics.Drawable.EnsureMutationAllowed(String action)
   at osu.Framework.Graphics.Transforms.Transformable.AddTransform(Transform transform, Nullable`1 customTransformID)
   at osu.Framework.Graphics.TransformableExtensions.TransformTo[TThis](TThis t, Transform transform)
   at osu.Framework.Graphics.TransformableExtensions.FadeOut[T](T drawable, Double duration, Easing easing)
   at osu.Game.Graphics.UserInterface.LoadingLayer.PopOut() in /home/dachb/Documents/opensource/osu/osu.Game/Graphics/UserInterface/LoadingLayer.cs:line 85
   at osu.Framework.Bindables.Bindable`1.TriggerValueChange(T previousValue, Bindable`1 source, Boolean propagateToBindings, Boolean bypassChecks)
   at osu.Framework.Bindables.Bindable`1.set_Value(T value)
   at osu.Game.Screens.Ranking.FavouriteButton.<getBeatmapSet>b__17_1(Exception e) in /home/dachb/Documents/opensource/osu/osu.Game/Screens/Ranking/FavouriteButton.cs:line 81
   at osu.Game.Online.API.APIRequest.TriggerFailure(Exception e) in /home/dachb/Documents/opensource/osu/osu.Game/Online/API/APIRequest.cs:line 181
   at osu.Game.Online.API.APIRequest.Fail(Exception e) in /home/dachb/Documents/opensource/osu/osu.Game/Online/API/APIRequest.cs:line 219
```

Layendan commented:
> Got this crash when testing the log in/out logic of the favourite button:

What platform are you testing on? And what steps are you taking to get the error? I can't seem to replicate the issue you're having.

Layendan committed:
Fix crash if favourite button api request fails

bdach committed:
Fix the fix

The more proper way to do this would be to address the underlying issue,
which is https://github.com/ppy/osu/issues/29546, but let's do this
locally for now.

bdach reviewed (approved):
probably fine

peppy reviewed (approved):
Very useful buttons to have, zallius will be happy

cdwcgt commented:
![image](https://github.com/user-attachments/assets/436cae1b-8357-4504-aa54-38d5a1e9c955)
I thought just dimming the button would be enough?
The notification pop-up here is annoying

Especially he will do this to the local beatmap.

bdach commented:
@cdwcgt i have no idea what i'm looking at on your screenshots or how you got that to happen. maybe use words to describe that if you want something to be done about it

cdwcgt commented:
if an beatmap is local beatmap or cannot fetch online, the `FavouriteButton` will pop an error notification *every* times I open result screen.
https://github.com/ppy/osu/blob/28d0a245556e3be98ad2d3612358d86ead9e0e27/osu.Game/Screens/Ranking/FavouriteButton.cs#L112-L121

you can check out this video

https://github.com/user-attachments/assets/30f56dba-4975-4300-a32a-5a48eedfd56b

bdach commented:
@cdwcgt i tested with several local unsubmitted beatmaps, as well as several local online-but-modified beatmaps, but cannot see the error notification you mention. please provide a better reproduction scenario

bdach commented:
well okay i got it to happen _but_ on a beatmap set that exists on `osu.ppy.sh` but does not exist on `dev.ppy.sh`. which is not at all what you were claiming.

cdwcgt commented:
This should be related to the onlineID marked in the beatmap set, In stable, map that have entered the upload page but have not uploaded will be assigned a beatmapsetID,
This map may have received an invalid mapset id at that time, and import to lazer.
This is what caused me to wonder, because I am sure my map was not uploaded and it has this problem on any server.

It can also reproduce by some map that have been deleted online but are still stored locally (such as some mania dan map?)
Just like you can right-click on a mapset to get details but you can't actually get the content.

this is the mapset file I use in video
[AOA - Xin Zuo Mi Ma (cdwcgt).zip](https://github.com/user-attachments/files/16906205/AOA.-.Xin.Zuo.Mi.Ma.cdwcgt.zip)

I have overlooked that my beatmaps may have beatmapsetID even though I didn't upload them.
